### ISSUE TITLE
Consider alternative syntax / signature for bracket

### ISSUE THREAD

**Issue by pchlupacek** [functional-streams-for-scala/fs2#666]
Currently onFinalize guaranteed to be executed when the process terminates irrelevant if it terminated with error or normally. 

This adds ability to cleanup code to be notified if the Stream terminated normally or failed. 

Same shall be added to Stream.bracket, which I would  propose in addition to have signature :

```scala 
def bracket[F[_],R,A](acquire: F[R],release: R => Option[Throwable] => F[Unit])(use: R => Stream[F,A]): Stream[F,A] =  ???
```

Note the swap of user and release and currying. This allows then to this syntax : 

```scala 

Stream.bracket(acquire,release) { resource => 
  // build the stream with resource. 
}


```

@pchiusano @mpilquist any ideas?

**Comment by pchiusano** [comment]
A couple points:

1. If we were to do this, I'd introduce `bracketWithCause` or something like that, not alter the signature of `bracket` or `onFinalize`.
2. But I don't think we should do this. Reason is we can't provide any sort of guarantees. A stream can terminate for three reasons: input exhaustion, an uncaught exception, or early termination of the `Pull` that is reading from it (which might be because of an uncaught exception in some other stream the `Pull` is reading from). Early termination can occur at the exact same moment as either input exhaustion or failure due to an uncaught exception (imagine: we've just reached the end of the stream and there are no more elements, but at that exact moment early termination occurs), so what we signal as 'the cause' would be nondeterministic, not something the user could rely upon - if we give them a combinator like this, it could be a nondeterministic bug waiting to happen.
3. You can always use `onError` on the stream you pass to `onFinalize` or in the stream that uses the resource in `bracket`. I feel this covers vast majority of use cases.
4. If you really want to write a combinator like this, I think you could do it with `onError` and `bracket`, and an `AtomicReference[Cause]`. It will have nondeterministic behavior. But let's not endorse people actually using such a thing.

So, not in favor. Sorry @pchlupacek :)

**Comment by pchlupacek** [comment]
@pchiusano I see you points. As regards (2) I think it can be nondeterministic only when having nondet combinator between streams right? I am not sure otherwise how this can be nondeterministic. An if yes, what actually onError is doing to avoid be subject to this race as well?

As far as the syntax goes, what are the reasoning behind current bracket syntax? I found it personally extremely unpleasant to work with....

**Comment by mpilquist** [comment]
@pchlupacek I agree that the current signature of `bracket` is rather annoying to work with. I do like it better than your proposed signature though, because with the current signature, `R` can be inferred. E.g., consider:

```
  // Current API
  Stream.bracket(acquire)(r => use(r), _.close)
  // Proposed
  Stream.bracket(acquire, _.close)(r => use(r)) // Fails to infer
```

If we want to improve the bracket syntax, I'd also consider the following alternatives. 

```
  def bracket[F[_],R,A](acquire: F[R])(use: R => Stream[F, A])(release: R => F[Unit]): Stream[F,A]
  def bracket[F[_],R,A](acquire: F[R])(f: R => (Stream[F, A], F[Unit])): Stream[F,A]
```

I personally find the last option the most intuitive but I'm also fine with the way it is now. I tend to do things like:

```
Stream.bracket(acquire)(
  { case ... },
  { case ... })
```

**Comment by pchiusano** [comment]
@pchlupacek yes you are right on 2) if your stream is totally deterministic. onError doesn't provide any guarantees - it can be interrupted midway through and shouldn't be relied on to do cleanup or anything important, it's more useful for making a best effort to recover from errors. bracket is the only combinator that limits when interruption can occur, everything else can be interrupted whenever, though we will never interrupt a user-provided effect midway through and have no means of doing that anyway.

I guess I could see the argument that, well, onError doesn't provide any guarantees, so why not add this other combinator that also doesn't provide any guarantees? Somewhat subjective but I just don't like that.

Re: syntax, I like it curried for inference reasons like @mpilquist says but other than that don't really care. I guess I haven't had a problem with the current syntax though.

**Comment by pchlupacek** [comment]
@pchiusano glad I understood it well, was sorta lost if I didn't miss something important in new implementation.  

I came across need for this when you want to have action that shall be executed when your stream terminates and shall depends for example on logging. then you have to do something like s.append().onError() which I just found boilerplate, and unnecessary split of functionality. 

Other that that I agree with what you say. However logging cause of termination in streams is important, and the fact that we couldn't guarantee that shall be probably very clearly documented in onError as well.

I'll perhaps experiment with different trade-offs of bracket and perhaps will come with some curried version (as current version is sort of half-step curried)

**Comment by pchiusano** [comment]
Cool. My feeling is still lets hold off on adding it to core. You can add it in your own code and we can revisit if there is a lot of demand for something like this.

Agree on beefing up docs, actually I think guide could use another appendix explaining subtleties of interruption and how this can bite you if you aren't using bracket to do resource cleanup.

**Comment by pchiusano** [comment]
We can keep this open to track any syntax changes for bracket. I'll open separate ticket for doc improvements.

**Comment by pchlupacek** [comment]
ok fine with that. Also, because of this Stream.attempt perhaps will need comment too. 

You know, I am just curious if we don't overeducating here. Isn't it obvious that if you do concurrency then all the error handling is nondet by nature?

**Comment by pchiusano** [comment]
I mean, maybe, but I think it is worth mentioning, especially since
`onComplete` used to be the recommended way of doing resource cleanup. I
have been tempted to just get rid of `onComplete` or at least rename it so
people are forced to evaluate each use site (perhaps `onFinish`?). I don't
have strong feelings on that.

I don't care too much about hunting down every error-handling combinator
and putting huge warnings in the scaladoc, but if you see any functions
that you think would benefit from better scaladoc, feel free to update.

At some point, you have to let users be responsible for reading the docs.
:) The migration guide is pretty clear on this point also: "All resources
should be acquired using bracket. Placing finalizers / cleanup actions in an
 onComplete will not guarantee they are run if a stream is being consumed
asynchronously or is terminated early by its consumer."

**Comment by mpilquist** [comment]
@pchlupacek Any more thoughts on `bracket` syntax?

**Comment by pchlupacek** [comment]
I am sort of leaning towards getting rid of onComplete all together. If we don't give any guarantees on it, then is so much cleaner to do .append().onError(). I cannot think for any use for onComplete except for resource cleanup which is not supposed to be the case now. So imho is better to give rid of it.

**Comment by mpilquist** [comment]
@pchlupacek I think I'm on board with getting rid of `onComplete`. Are you OK with the signature of `bracket` as-is then?

**Comment by pchlupacek** [comment]
@mpilquist I can live with that :-)

**Comment by pchlupacek** [comment]
@mpilquist I can likely do this till EOW

**Comment by mpilquist** [comment]
@pchlupacek No worries -- I removed `onComplete` in #707. If that looks good to you and you are happy with the signature of `bracket`, then let's get it merged and release RC1. :)

**Comment by mpilquist** [comment]
Per discussion in Gitter, this can be closed now that #707 is merged.
